### DISCUSSION TITLE
Best-fit companion library to Refit on the server side?

### DISCUSSION THREAD

**Discussion by bancroftway** [reactiveui/refit#1776] [Q&A] [answered]
While Refit is very helpful for type-safe api client, is there a server api helper library that would complement Refit very well?

As an example, I have been looking at Fastendpoints, however, in that library, it would still require setting up the endpoint route and http verb again in the Configure method. I am really hoping that there is an elegant way to write the api Interface once in a Shared library (such as in a ASP.Net Core hosted Blazor Webassembly project), and then be able to use that api interface not only on the client side in refit but also the server.

**Comment by glennawatson** [comment, marked as answer]
There is no official Refit server-side counterpart, but a couple of community approaches fit the "define the interface once, share it" goal:

- Refitter (https://github.com/christianhelle/refitter) generates Refit clients from OpenAPI, but it is still client-side.
- For sharing one interface between an ASP.NET Core server and a Refit client, look at "shared contract" libraries that bind ASP.NET Core endpoints to an interface, for example:
  - SignalR-style or gRPC contract-first if you can move off REST.
  - Community projects that map controller routes from an interface (search for "ASP.NET Core interface to endpoint" / "contract-first REST"). These let the interface drive both sides.

The reason this is not built into Refit: Refit only models the client. Mapping the same interface to server routing requires an ASP.NET Core integration, which is out of scope for this library.

If you want a single source of truth, the most robust paths today are gRPC (contract-first by design) or generating both client and server from an OpenAPI spec. For pure REST with a shared interface, you will need a server-side route-mapping helper; FastEndpoints does not currently bind from an interface.
